### PR TITLE
Solver factory shouldn't have its own namespacet instance

### PR DESCRIPTION
**Pull request by peterschrammel**
A reference of the namespace is passed on into the solver allocated.
If the solver factory goes out of scope then the solver will crash at unexpected places.

This PR doesn't propose a safe solution, but makes the namespace instance visible for the caller.

The underlying problem is that there is no clear ownership of namespacet objects. They are wildly passed around and re-instantiated throughout the code base. 
Without having a clear ownership convention, we should at least use `shared_ptr` to be safe.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: cb184831119c6741b4ceba7f399ccea225fc6052).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97472916
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by smowton — APPROVED** [review]

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
I think we should mostly stop passing around `namespacet` objects - they are very cheap to create; instead the symbol table (or multiple symbol tables) should be passed around.

**Comment by peterschrammel** [comment]
Waiting for bump to succeed before merge.

**Pull request merged**
